Sources/BrowserServicesKit/FeatureFlagger/FeatureFlagger.swift

```diff
@@ -0,0 +1,115 @@
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+public protocol FeatureFlagger {
+
+/// Called from app features to determine whether a given feature is enabled.
+///
+/// `forProvider: F` takes a FeatureFlag type defined by the respective app which defines from what source it should be toggled
+/// see `FeatureFlagSourceProviding` comments below for more details
+    func isFeatureOn<F: FeatureFlagSourceProviding>(forProvider: F) -> Bool
+ }
+
+public class DefaultFeatureFlagger: FeatureFlagger {
+    private let internalUserDecider: InternalUserDecider
+    private let privacyConfig: PrivacyConfiguration
+
+    public init(internalUserDecider: InternalUserDecider, privacyConfig: PrivacyConfiguration) {
+        self.internalUserDecider = internalUserDecider
+        self.privacyConfig = privacyConfig
+    }
+
+    public func isFeatureOn<F: FeatureFlagSourceProviding>(forProvider provider: F) -> Bool {
+        switch provider.source {
+        case .disabled:
+            return false
+        case .internalOnly:
+            return internalUserDecider.isInternalUser
+        case .remoteDevelopment(let featureType):
+            guard internalUserDecider.isInternalUser else {
+                return false
+            }
+            return isEnabled(featureType)
+        case .remoteReleasable(let featureType):
+            return isEnabled(featureType)
+        }
+    }
+
+    private func isEnabled(_ featureType: PrivacyConfigFeatureLevel) -> Bool {
+        switch featureType {
+        case .feature(let feature):
+            return privacyConfig.isEnabled(featureKey: feature)
+        case .subfeature(let subfeature):
+            return privacyConfig.isSubfeatureEnabled(subfeature)
+        }
+    }
+}
+
+
+/// To be implemented by the FeatureFlag enum type in the respective app. The source corresponds to
+/// where the final value should come from.
+///
+/// Example:
+//
+// public enum FeatureFlag: FeatureFlagSourceProviding {
+//    case sync
+//    case autofill
+//    case cookieConsent
+//    case duckPlayer
+//
+//    var source: FeatureFlagSource {
+//        case .sync:
+//            return .disabled
+//        case .cookieConsent:
+//            return .internalOnly
+//        case .credentialsAutofill:
+//            return .remoteDevelopment(.subfeature(AutofillSubfeature.credentialsAutofill))
+//        case .duckPlayer:
+//            return .remoteReleasable(.feature(.duckPlayer))
+//    }
+// }
+
+
```

Comment:
I noticed that Xcode's doc popover wasn't displaying this comment when option-clicking `FeatureFlagSourceProviding`, I think removing these spaces should fix that.